### PR TITLE
Allow unique-callout module to be dis/en-abled using cargo flags

### PR DESCRIPTION
**Pull request by rahulanand16nov**
As previously discussed, even though unique callouts perform better in accuracy and performance, there is still doubt about its reliability. So, it's best to allow users to enable or disable this module using cargo feature flags.

To minimize the use of `cfg` all around the code (it's an anti-pattern to do so), I have added dead code in the `unique_callout.rs` to have them as placeholders if the module is disabled.

After this PR is merged, I'll create a new PR to merge this branch into main to squash all the commits into one merge commit.

### CODE REVIEW

**Comment by rahulanand16nov** [comment]
All the CI checks are working fine locally but if someone is still unsure, note that merge PR will run the CI regardless ;)

**Review by NomadXD — APPROVED** [review]
LGTM

**Review by unleashed — CHANGES_REQUESTED** [review]
Yes, it's a total antipattern - we don't want to constantly add `#[cfg(feature = "unique_callout")]` everywhere. What we want instead is 1-2 such instances in which one of those is used to include the relevant code in a module, possibly with a `#[cfg(not(feature = "unique_callout"))]` including a dummy module with the same interface that does nothing (other options are possible), and then perhaps, very sparingly and with good justification, use it elsewhere to enable/disable other code.

**Comment by rahulanand16nov** [comment]
@unleashed Added a dummy module with the reasoning behind it as well. I think this is the minimum number of `cfg` attributes that are needed and cannot decrease the count more due to coupling between the `unique_callout` and other modules.

File: cache-filter/src/filter/http.rs [hunk 1]
```diff
@@ -1,9 +1,12 @@
+#[cfg(not(feature = "unique_callout"))]
+use crate::dummy::*;
+#[cfg(feature = "unique_callout")]
+use crate::unique_callout::{
+    free_callout_lock_and_notify_waiters, set_callout_lock, SetCalloutLockStatus, WaiterAction,
+};
```
Comment (unleashed) [inline cache-filter/src/filter/http.rs:6]: I think a better solution could be: (not checked, maybe this ends up being not workable)

```rust
#[cfg(not(feature = "unique_callout"))]
use crate::unique_callout_dummy as unique_callout;
#[cfg(feature = "unique_callout")]
use crate::unique_callout;

use unique_callout::{free_callout_lock_and_notify_waiters, set_callout_lock, SetCalloutLockStatus, WaiterAction};
```

File: cache-filter/src/filter/root.rs [hunk 2]
```diff
@@ -1,18 +1,22 @@
...
+// Following imports are only needed when unqiue-callout feature is enabled.
+#[cfg(feature = "unique_callout")]
+use {
+    crate::unique_callout::{WaiterAction, WAITING_CONTEXTS},
+    crate::utils::request_process_failure,
+    proxy_wasm::hostcalls::set_effective_context,
+    threescale::{
+        proxy::{get_app_id_from_cache, get_application_from_cache},
+        structs::AppIdentifier,
+    },
```
Comment (unleashed) [inline cache-filter/src/filter/root.rs:19]: If this is only needed for `on_queue_ready`, then we should `use` them only within the body of `on_queue_ready`.

File: cache-filter/src/lib.rs [hunk 3]
```diff
@@ -3,8 +3,11 @@
 const VM_ID: &str = "my_vm_id";
 
 mod configuration;
+#[cfg(not(feature = "unique_callout"))]
+mod dummy;
 mod filter;
 mod log;
 mod rand;
+#[cfg(feature = "unique_callout")]
```
Comment (unleashed) [inline cache-filter/src/lib.rs:11]: Renaming `dummy` to `unique_callout_dummy` will help place these two together.

File: cache-filter/src/unique_callout.rs [hunk 4]
```diff
@@ -60,6 +60,7 @@ pub enum WaiterAction {
 }
 
 // This enum is used to give out status for an http context trying to get callout-lock.
+#[cfg(feature = "unique_callout")]
```
Comment (unleashed) [inline cache-filter/src/unique_callout.rs:63]: Isn't this module already only compiled in when this is true?

File: cache-filter/src/filter/http.rs [hunk 5]
```diff
@@ -1,9 +1,12 @@
+#[cfg(not(feature = "unique_callout"))]
+use crate::dummy::*;
```
[the remark this answers is not in the document]
Comment (rahulanand16nov) [reply cache-filter/src/filter/http.rs:2]: I was thinking of having a single dummy for `unqiue_callout` and `visible_logs` but `unique_callout_dummy` is also fine.

File: cache-filter/src/filter/root.rs (hunk 2 above)
Comment (rahulanand16nov) [reply cache-filter/src/filter/root.rs:19]: Way better!

File: cache-filter/src/unique_callout.rs (hunk 4 above)
Comment (rahulanand16nov) [reply cache-filter/src/unique_callout.rs:63]: Forgot to remove it :weary:

File: cache-filter/src/filter/http.rs (hunk 1 above)
Comment (rahulanand16nov) [reply cache-filter/src/filter/http.rs:6]: Workable!

File: cache-filter/src/lib.rs (hunk 3 above)
Comment (rahulanand16nov) [reply cache-filter/src/lib.rs]: Done

**Comment by rahulanand16nov** [comment]
Rebased and CI is passing locally.

**Review by unleashed — COMMENTED** [review]
Ok @rahulanand16nov - now let's squash the reverts and the rename of the module with each diff belonging to each one of the relevant commits so the history is kept nice and clean!

File: cache-filter/Cargo.toml [hunk 6]
```diff
@@ -16,6 +16,7 @@ crate-type = ["cdylib"]
 [features]
 default = ["prng_pcg32"]
 visible_logs = []
+unique_callout = []
```
Comment (unleashed) [inline cache-filter/Cargo.toml:19]: nitpick: typo in the commit message

Comment (rahulanand16nov) [reply cache-filter/Cargo.toml:19]: Fixed :D

**Comment by rahulanand16nov** [comment]
@unleashed LGTM now, please approve and merge if it looks good to you as well :)

**Review by unleashed — APPROVED** [review]

**Comment by unleashed** [comment]
Looks good!

**Pull request merged**